web/web.go

```diff
@@ -0,0 +1,694 @@
+	bufWriter.Flush()
+	overview := buf.String()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(overview)
+}
+
+func getTypeMap(w http.ResponseWriter, r *http.Request) {
+	if app.conv == nil || app.driver == "" {
+		http.Error(w, fmt.Sprintf("Schema is not converted or Driver is not configured properly. Please retry converting the database to spanner."), http.StatusNotFound)
+		return
+	}
+	var typeMap map[string][]typeIssue
+	switch app.driver {
+	case "mysql", "mysqldump":
+		typeMap = mysqlTypeMap
+	case "postgres", "pg_dump":
+		typeMap = postgresTypeMap
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+		return
+	}
+	// Filter typeMap so it contains just the types SrcSchema uses.
+	filteredTypeMap := make(map[string][]typeIssue)
+	for _, srcTable := range app.conv.SrcSchema {
+		for _, colDef := range srcTable.ColDefs {
+			if _, ok := filteredTypeMap[colDef.Type.Name]; ok {
+				continue
+			}
+			filteredTypeMap[colDef.Type.Name] = typeMap[colDef.Type.Name]
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(filteredTypeMap)
+}
+
+func setTypeMapGlobal(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var t map[string]string
+	err = json.Unmarshal(reqBody, &t)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+
+	for k, v := range app.conv.SpSchema {
+		for kk, _ := range v.ColDefs {
+			for tk, tv := range t {
+				sourceTable := app.conv.ToSource[k].Name
+				sourceCol := app.conv.ToSource[k].Cols[kk]
+				srcCol := app.conv.SrcSchema[sourceTable].ColDefs[sourceCol]
+				if srcCol.Type.Name == tk {
+					var ty ddl.Type
+					var issues []internal.SchemaIssue
+					switch app.driver {
+					case "mysql", "mysqldump":
+						ty, issues = toSpannerTypeMySQL(srcCol.Type.Name, tv, srcCol.Type.Mods)
+					case "pg_dump", "postgres":
+						ty, issues = toSpannerTypePostgres(srcCol.Type.Name, tv, srcCol.Type.Mods)
+					default:
+						http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+						return
+					}
+					if len(srcCol.Type.ArrayBounds) > 1 {
+						ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+						issues = append(issues, internal.MultiDimensionalArray)
+					}
+					if srcCol.Ignored.Default {
+						issues = append(issues, internal.DefaultValue)
+					}
+					if srcCol.Ignored.AutoIncrement {
+						issues = append(issues, internal.AutoIncrement)
+					}
+					if len(issues) > 0 {
+						app.conv.Issues[sourceTable][srcCol.Name] = issues
+					}
+					ty.IsArray = len(srcCol.Type.ArrayBounds) == 1
+					spSchema := app.conv.SpSchema[k]
+					colDef := spSchema.ColDefs[kk]
+					colDef.T = ty
+					spSchema.ColDefs[kk] = colDef
+					app.conv.SpSchema[k] = spSchema
```

Comment:
This last assignment isn't needed because we've only updated the map part of SpSchema and maps are actually pointers. So the previous assignment to spSchema.ColDefs has already updated the map in app.conv.SpSchema[k].

Reply:
removed last assignment